client/directclient.go

```diff
@@ -80,58 +82,170 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
+	return len(q.writers) >= q.q && len(q.responses)+len(q.writers) >= q.replicaCount
+}
+
+func (q *stdQuorumer) readyWriters() (wrs []io.Writer) {
+	for _, w := range q.writers {
+		wrs = append(wrs, w)
+	}
+	return
+}
+
+func (q *stdQuorumer) closeWriters() {
+	for _, w := range q.writers {
+		w.Close()
 	}
-	quorum := int(math.Ceil(float64(len(reqs)) / 2.0))
-	responseClasses := []int{0, 0, 0, 0, 0, 0}
-	responseCount := 0
-	var chosenResponse *http.Response
+}
+
+func newQuorumer(r ring.Ring, partition uint64, makeRequest func(index int, dev *ring.Device)) quorumer {
+	return &stdQuorumer{
+		launchRequest: makeRequest,
+		writers:       nil,
+		q:             int(math.Ceil(float64(r.ReplicaCount()) / 2.0)),
+		requests:      0,
+		requestCount:  0,
+		replicaCount:  int(r.ReplicaCount()),
+		responses:     nil,
+		failures:      nil,
+		devs:          r.GetNodes(partition),
+		more:          r.GetMoreNodes(partition),
+		rc:            make([]int, 5),
+	}
+}
+
+// quorumResponse returns with a response representative of a quorum of nodes.
+//
+// This is analogous to swift's best_response function.
+func (c *ProxyDirectClient) quorumResponse(r ring.Ring, partition uint64, devToRequest func(int, *ring.Device) (*http.Request, error)) (chosenResponse *http.Response) {
+	type respRec struct {
+		i int
+		r *http.Response
+	}
+	responses := make(chan *respRec)
+	q := newQuorumer(r, partition, func(index int, dev *ring.Device) {
+		if req, err := devToRequest(index, dev); err != nil {
+			responses <- &respRec{i: index, r: ResponseStub(http.StatusInternalServerError, err.Error())}
+		} else if r, err := c.client.Do(req); err != nil {
+			responses <- &respRec{i: index, r: ResponseStub(http.StatusInternalServerError, err.Error())}
+		} else {
+			responses <- &respRec{i: index, r: StubResponse(r)}
+		}
+	})
+	q.start()
 	for response := range responses {
-		responseCount++
-		class := response.StatusCode / 100
-		if class <= 5 {
-			responseClasses[class]++
-			if responseClasses[class] >= quorum {
-				chosenResponse = response
-				break
-			}
+		var met bool
+		q.addResponse(response.i, response.r)
+		if met, chosenResponse = q.quorum(); met {
+			break
+		} else if !q.quorumPossible() {
+			return ResponseStub(http.StatusServiceUnavailable, "The service is currently unavailable.")
 		}
 	}
-	// Give any pending requests *some* chance to finish. This will increase
-	// the likelihood that a read immediately after a write will get the latest
-	// information, as we'll force the caller to wait until all backends have
-	// responded, with a timeout limit.
+	// Give any pending requests *some* chance to finish.
 	timeout := time.After(PostQuorumTimeoutMs * time.Millisecond)
-waiting:
-	for responseCount < len(reqs) {
+	for q.outstanding() {
 		select {
-		case <-responses:
-			responseCount++
+		case response := <-responses:
+			q.addResponse(response.i, response.r)
 		case <-timeout:
-			break waiting
+			return
```

Comment:
Go back to the loop?  No, it was breaking out of the loop to hit the return below.  But just returning here let me get rid of the loop label.